src/lightmap.cpp

```diff
@@ -124,44 +125,33 @@ bool map::build_transparency_cache( const int zlev )
 
                     for( const auto &fld : cur_submap->fld[sx][sy] ) {
                         const field_entry &cur = fld.second;
-                        const field_id type = cur.get_field_type();
-                        const int density = cur.get_field_intensity();
+                        const field_type_id type = cur.get_type();
+                        const int density = cur.get_density();
 
-                        if( all_field_types_enum_list[type].transparent[density - 1] ) {
+                        if( type.obj().transparent( density - 1 ) ) {
                             continue;
                         }
 
                         // Fields are either transparent or not, however we want some to be translucent
-                        switch( type ) {
-                            case fd_cigsmoke:
-                            case fd_weedsmoke:
-                            case fd_cracksmoke:
-                            case fd_methsmoke:
-                            case fd_relax_gas:
-                                value *= 5;
-                                break;
-                            case fd_smoke:
-                            case fd_incendiary:
-                            case fd_toxic_gas:
-                            case fd_tear_gas:
-                                if( density == 3 ) {
-                                    value = LIGHT_TRANSPARENCY_SOLID;
-                                } else if( density == 2 ) {
-                                    value *= 10;
-                                }
-                                break;
-                            case fd_nuke_gas:
-                                value *= 10;
-                                break;
-                            case fd_fire:
-                                value *= 1.0 - ( density * 0.3 );
-                                break;
-                            default:
+                        if( type == fd_cigsmoke || type == fd_weedsmoke || type == fd_cracksmoke ||
+                            type == fd_methsmoke || type == fd_relax_gas ) {
+                            value *= 5;
+                        } else if( type == fd_smoke || type == fd_incendiary || type == fd_toxic_gas ||
+                                   type == fd_tear_gas ) {
+                            if( density == 3 ) {
                                 value = LIGHT_TRANSPARENCY_SOLID;
-                                break;
+                            } else if( density == 2 ) {
+                                value *= 10;
+                            }
+                        } else if( type == fd_nuke_gas ) {
+                            value *= 10;
+                        } else if( type == fd_fire ) {
+                            value *= 1.0 - ( density * 0.3 );
+                        } else {
+                            value = LIGHT_TRANSPARENCY_SOLID;
```

Comment:
Was there something wrong with the switch case?  switch is way easier to grok than a bunch of if's

Reply:
Not to mention `switch` is generally more performant than `if/else`

Reply:
Unfortunately switch only works with basic types, which `int_id<T>` is not. I have some thoughts on fixing this monster function later anyway.